### PR TITLE
linker: aarch32: simplify memory region generation

### PR DESCRIPTION
**Pull request by JordanYates**
Generating linker memory regions from devicetree nodes is a common pattern in the linker script. This PR adds a macro which automatically generates a region given a devicetree node. The macro is then used in the linker script to cleanup definitions.

Dedicated devicetree nodes are created for SRAM1 and SRAM2  for the STM32WB family to facilitate macro usage.
The TI_CCFG registers are redefined as a partition on `flash0` to enable more cleanup.

### CODE REVIEW

**Comment by JordanYates** [comment]
@galak This contains the standalone linker cleanup from #33656

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld [hunk 1]
```diff
@@ -107,34 +86,20 @@ _region_min_align = 4;
...
-#endif
-#ifdef CONFIG_STM32_BACKUP_SRAM
-    BACKUP_SRAM           (rw) : ORIGIN = DT_REG_ADDR(DT_NODELABEL(backup_sram)), LENGTH = DT_REG_SIZE(DT_NODELABEL(backup_sram))
-#endif
+    /* TI CCFG Registers */
+    DT_REGION_FROM_NODE_STATUS_OKAY(FLASH_CCFG, rwx, DT_NODELABEL(ti_ccfg_partition))
+    /* STM32 memory regions */
+    DT_REGION_FROM_NODE_STATUS_OKAY(CCM, rw, DT_CHOSEN(zephyr_ccm))
+    DT_REGION_FROM_NODE_STATUS_OKAY(ITCM, rw, DT_CHOSEN(zephyr_itcm))
+    DT_REGION_FROM_NODE_STATUS_OKAY(DTCM, rw, DT_CHOSEN(zephyr_dtcm))
```
Comment (erwango) [inline include/arch/arm/aarch32/cortex_m/scripts/linker.ld:95]: ccm is indeed STM32 specific but itcm and dtcm are arm generic.

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld [hunk 2]
```diff
@@ -107,34 +86,20 @@ _region_min_align = 4;
...
+    /* STM3232WB IPC RAM */
```
Comment (erwango) [inline include/arch/arm/aarch32/cortex_m/scripts/linker.ld:96]:
```suggestion
    /* STM32WB IPC RAM */
```

**Review by erwango — CHANGES_REQUESTED** [review]
Minor changes requested but otherwise welcomed addition, thanks!

**Review by erwango — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]
Nice cleanup, thanks. :+1:

**Review by MaureenHelm — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by ioannisg** [comment]
Nice work here - sorry I had no chance to review this.

**Comment by MaureenHelm** [comment]
Turns out this broke code relocation to these regions, as the relocation script expects `<region name>_ADDR` to be defined (i.e., `ITCM_ADDR`). This affects #33701, which relocates flash driver code to ITCM.

cc: @pdgendt @mmahadevan108

**Comment by JordanYates** [comment]
@MaureenHelm is there a sample which produces the error, or can be modified to produce it?
I'm not sure which relocation script is the problem, nor can I find where `<region name>_ADDR` is expected.

**Comment by pdgendt** [comment]
@JordanYates the [relocate script](https://github.com/zephyrproject-rtos/zephyr/blob/d8bbd2ad5c57de7ad85b920477447791518725c2/scripts/gen_relocate_app.py#L63) creates the `<region name>_ADDR`

**Comment by JordanYates** [comment]
@pdgendt thanks for that link, can you provide a board & sample which fails to build?

**Comment by pdgendt** [comment]
@JordanYates you can build `samples/drivers/flash_shell/` for `mimxrt1064_evk` but you have to enable the following in `prj.conf`:
```
CONFIG_FLASH_MCUX_FLEXSPI_NOR=y
CONFIG_FLASH_MCUX_FLEXSPI_XIP=y
```

**Comment by JordanYates** [comment]
@pdgendt  #34331 should fix the issue.

**Pull request merged**